File: 622-design-circular-queue/622-design-circular-queue.cpp
class MyCircularQueue {
public:
    int arr[1000] = {0};
    int k=0;
    int front = -1;
    int rear = -1;
    MyCircularQueue(int k) {
      this->k = k;
    }
    bool enQueue(int value) {
      if(this->isFull())
        return false;
      if(this->isEmpty())
        this->front = 0;
      this->rear = (this->rear+1)%k;
      this->arr[this->rear]=value;
      return true;
    }
    bool deQueue() {
      if(this->isEmpty())
        return false;
      if(this->front == this->rear){
        this->front = -1;
        this->rear = -1;
        return true;
      }
      this->front = (front+1)%this->k;
      return true;
    }
    int Front() {
      if(this->isEmpty())
        return -1;
      return this->arr[this->front];
    }
    int Rear() {
      if(this->isEmpty())
        return -1;
      return this->arr[this->rear];
    }
    bool isEmpty() {
      return this->front == -1 && this->rear == -1;
    }
    bool isFull() {
      return (this->rear+1)%this->k == this->front;
    }
};

File: 622-design-circular-queue/622-design-circular-queue.py
class MyCircularQueue:
    def __init__(self, k: int):
      self.front = -1
      self.back = -1
      self.cap = k
      self.arr = [0]*self.cap
    def enQueue(self, value: int) -> bool:
      if(self.isFull()):
        return False;
      if(self.isEmpty()):
        self.front = 0
      self.back = (self.back+1)%self.cap
      self.arr[self.back] = value
      return True;
    def deQueue(self) -> bool:
      if(self.isEmpty()):
        return False;
      if(self.front == self.back):
        self.front = self.back = -1
        return True
      self.front = (self.front+1)%(self.cap)
      return True;
    def Front(self) -> int:
      if(self.isEmpty()):
        return -1
      return self.arr[self.front]
    def Rear(self) -> int:
      if(self.isEmpty()):
        return -1
      return self.arr[self.back]
    def isEmpty(self) -> bool:
      return self.front == self.back == -1
    def isFull(self) -> bool:
      return (self.back+1)%self.cap == self.front
